[pulsar-functions] Fix message consume ordering issue of function

Tiscs opened the pull request:
Fix #7707
Fix message consume ordering issue of function with EFFECTIVELY_ONCE processing guarantees.


### Motivation

Since v2.5.2, function with EFFECTIVELY_ONCE processing guarantees consume message out of ordering when fail, because process a CompletableFuture result by async.

### Modifications

- Get result from future synchronized when processing guarantees of java instance equals `EFFECTIVELY_ONCE`.

jiazhai commented:
/pulsarbot run-failure-checks

Tiscs commented:
Rebased from `origin/master`.

sijie commented:
@Tiscs Can you rebase the pull request to latest master again?

Tiscs commented:
> @Tiscs Can you rebase the pull request to latest master again?

@sijie Done.

Tiscs commented on pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:
```diff
@@ -415,28 +415,45 @@ private void setupStateTable() throws Exception {
-                srcRecord.fail();
+    private void processResult(Record srcRecord, JavaExecutionResult result, Throwable throwable) throws Exception {
+        throwable = throwable == null ? result.getUserException() : throwable;
+        if (result.getUserException() != null) {
```
Should replace `result.getUserException()` to local variable `throwable` here?

jiazhai commented:
@Tiscs Thanks for the contribution, it seems there are still some conflicts with the master. Would you please help rebase this PR again?

Tiscs commented:
> @Tiscs Thanks for the contribution, it seems there are still some conflicts with the master. Would you please help rebase this PR again?

Rebased again, and fixed exception detection in `processResult`, but some checks were failed.

jiazhai commented:
This pr seems brings too much un-related changes by the merge. Would you please follow [this guide](http://pulsar.apache.org/en/contributing/#clone-the-repository-locally) to provide a new PR?

Tiscs commented:
Closed by new PR #8026.
